### PR TITLE
Expand cases for generating a witness in prepare inputs 

### PR DESCRIPTION
**Pull request by vezenovm**
# Related issue(s)

Resolves issue with not being able to use inputs to gadgets that had been set through a conditional statement.

# Description

While writing the account circuit I was performing this conditional:
```
let mut signer: [Field; 2] = [signing_pub_key[0], signing_pub_key[1]];
    if create == 1 {
        signer[0] = account_public_key[0];
        signer[1] = account_public_key[1];
    }
```
Which was later used by this intrinsic gadget: 
```
 let sig_res = std::schnorr::verify_signature(signer[0], signer[1], signature, message_byte_array);
```
When compiling this would lead to the error below:

Screen Shot 2022-11-08 at 4 40 18 PM

We require that inputs to intrinsic gadgets be transformed into witnesses, however, we were not handling the case where the inputs to the function are a result of a conditional instruction. The result of this conditional does not have a witness associated with it.

## Summary of changes

I simply added extra logic to `prepare_inputs` for cases other than a `NodeObj::Obj`. In the case that the NodeId is in the arithmetic cache, but does not yet have a witness generate, we simply generate a new witness for this instruction with `generate_witness`. 

## Dependency additions / changes

(If applicable.)

## Test additions / changes

I added a little extra logic to our `scalar_mul` test. If you run `nargo compile` on this test using a nargo build from the current master you should see the error that I mentioned above. The test passes as expected with the current changes. You can also change the input of `a` for the `scalar_mul` test to see how the inputs to `std::fixed_based::scalar_mul` are changed. 

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [X] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.

# Additional context

(If applicable.)

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/node.rs [hunk 1]
```diff
@@ -716,7 +716,7 @@ impl Binary {
         Binary::new(operator, lhs, rhs)
     }
 
-    fn evaluate<F>(
+    pub fn evaluate<F>(
```
Comment (guipublic) [inline crates/noirc_evaluator/src/ssa/node.rs:719]: I believe you do not need to make it pub

File: crates/noirc_evaluator/src/ssa/acir_gen.rs [hunk 2]
```diff
@@ -526,10 +526,12 @@ impl Acir {
                 }
                 _ => {
                     if self.arith_cache.contains_key(a) {
-                        if let Some(w) = self.arith_cache[a].clone().witness {
+                        let var = self.arith_cache[a].clone();
+                        if let Some(w) = var.witness {
```
Comment (guipublic) [inline crates/noirc_evaluator/src/ssa/acir_gen.rs:530]: Not a big deal, but if you think the following is more readable:
```
let witness = if let Some(w) = var.witness { w} else {generate_witness(&var, evaluator)};
inputs.push(GadgetInput { witness, num_bits: l_obj.size_in_bits() });
```

File: crates/noirc_evaluator/src/ssa/acir_gen.rs [hunk 3]
```diff
@@ -501,7 +501,7 @@ impl Acir {
                                         inputs.push(GadgetInput { witness: wit, num_bits });
                                     } else {
                                         //TODO we should store the witnesses somewhere, else if the inputs are re-used
-                                        //we will duplicate the witnesses.
+                                        //we will duplic  ate the witnesses.
```
Comment (guipublic) [inline crates/noirc_evaluator/src/ssa/acir_gen.rs:504]: you can remove this change :)

**Review by guipublic — CHANGES_REQUESTED** [review]

File: crates/noirc_evaluator/src/ssa/acir_gen.rs (hunk 3 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:504]: Must have accidentally done that

File: crates/noirc_evaluator/src/ssa/node.rs (hunk 1 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/node.rs:719]: This was an old change I forgot to remove, done

File: crates/noirc_evaluator/src/ssa/acir_gen.rs (hunk 2 above)
Comment (vezenovm) [reply crates/noirc_evaluator/src/ssa/acir_gen.rs:530]: I changed to using `unwrap_or_else`

**Review by guipublic — APPROVED** [review]

**Pull request merged**